src/box_test.jl

```diff
@@ -0,0 +1,87 @@
+# box_test.jl
+# Box-Pierce and Ljung-Box tests for autocorrelation
+#
+# Copyright (C) 2017   Benjamin Born
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+export BoxPierceTest, LjungBoxTest
+
+# Box-Pierce test
+
+immutable BoxPierceTest <: HypothesisTest
+    y::Vector{Float64}  # time series vector
+    n::Int              # number of observations
+    lag::Int            # number of lags in test statistic
+    dof::Int            # degrees-of-freedom correction (if y is residual)
+    Q::Float64          # test statistic
+end
+
+function BoxPierceTest{T<:Real}(y::AbstractVector{T}, lag::Int, dof::Int=0)
+    if dof>=lag
+        throw(ArgumentError("Number of lags has to be larger than d-o-f correction"))
+    end
+	n = size(y)[1]
+	Q = n * mapreduce(k -> autocor(y,k:k).^2, +, 1:lag)[1]
```

Comment:
`n * first(sum(k -> autocor(y, k:k).^2, 1:lag))`

`first` is equivalent to `[1]` but I usually prefer it for its generality. Also, the `mapreduce` can just be `sum` with a function argument, since `sum(f, x)` just calls `mapreduce(f, +, x)`.